app/Http/Controllers/AdminTypesController.php

```diff
@@ -0,0 +1,128 @@
+     * Display a listing of the resource.
+     *
+     * @return \Illuminate\Http\Response
+     */
+    public function redirectWithMessage($routeName, $message)
+    {
+        return redirect()->route($routeName)->with('notice', trans($message));
+    }
+
+    public function index()
+    {
+        $type = Type::all();
+        return view('admin-layout.tour-types.type-list', compact('type'));
+    }
+
+    /**
+     * Show the form for creating a new resource.
+     *
+     * @return \Illuminate\Http\Response
+     */
+    public function create()
+    {
+        $type = Type::where('parent_id', null)->get();
+        return view('admin-layout.tour-types.type-add', compact('type'));
+    }
+
+    /**
+     * Store a newly created resource in storage.
+     *
+     * @param  \Illuminate\Http\Request  $request
+     * @return \Illuminate\Http\Response
+     */
+    public function store(TypesRequest $request)
+    {
+        $type = new Type;
+        $type->name = $request->name;
+        $type->parent_id = $request->parentType;
+        $type->save();
+
+        return ($this->redirectWithMessage('types.create', 'language.addMessage'));
+    }
+
+    /**
+     * Display the specified resource.
+     *
+     * @param  int  $id
+     * @return \Illuminate\Http\Response
+     */
+    public function show($id)
+    {
+        //
+    }
+
+    /**
+     * Show the form for editing the specified resource.
+     *
+     * @param  int  $id
+     * @return \Illuminate\Http\Response
+     */
+    public function edit($id)
+    {
+        $parentType = Type::where('parent_id', null)->get();
+        $type = Type::find($id);
+        if ($type == null)
+        {
+            return ($this->redirectWithMessage('types.index', 'language.addMessage'));
+        }
+        else
+        {
+            return view('admin-layout.tour-types.type-edit', compact('parentType', 'type'));
+        }
+    }
+
+    /**
+     * Update the specified resource in storage.
+     *
+     * @param  \Illuminate\Http\Request  $request
+     * @param  int  $id
+     * @return \Illuminate\Http\Response
+     */
+    public function update(TypesRequest $request, $id)
+    {
+        $type = Type::find($id);
+        if ($type == null)
+        {
+            return ($this->redirectWithMessage('types.index', 'language.notFoundMessage'));
+        }
+        else
+        {
+            $type->name = $request->name;
+            $type->parent_id = $request->parentType;
+            $type->save();
+
+            return ($this->redirectWithMessage('types.index', 'language.updateMessage'));
+        }
+    }
+
+    /**
+     * Remove the specified resource from storage.
+     *
+     * @param  int  $id
+     * @return \Illuminate\Http\Response
+     */
+    public function destroy($id)
+    {
+        $type = Type::find($id);
+        if ($type == null)
+        {
+            return ($this->redirectWithMessage('types.index', 'language.notFoundMessage'));
+        }
+        else
```

Comment:
đoạn code này lặp lại nhiều quá nhỉ? k tìm thấy thì đưa chung 1 message: k tìm thấy
viết riêng acction rồi dùng chung đi em

Reply:
Dạ anh

Reply:
em chưa fix mà. 
$type = Type::find($id);
        if ($type == null)
        {
            return ($this->redirectToIndexWithNotFoundMessage());
        }
đoạn code này lặp lại nhiều quá này e

Reply:
À dạ em nghĩ là chỉ cái lặp cái redirect ạ. Em đang fix lại đây ạ